### PR TITLE
Implement DeclareTypesInNamespaces

### PR DESCRIPTION
**Pull request by nguerrera**
Cherry-picking an easy one. ;) I think the fixer should be easy enough, but haven't done it yet. Will stop by team room for some guidance on that since I haven't touched fixers before.

#396 

/cc @srivatsn @lgolding

### CODE REVIEW

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/DeclareTypesInNamespaces.cs [hunk 1]
```diff
@@ -26,14 +28,26 @@ public abstract class DeclareTypesInNamespacesAnalyzer : DiagnosticAnalyzer
...
+
+        private void AnalyzeSymbol(SymbolAnalysisContext context)
+        {
+            ISymbol type = context.Symbol;
+
+            if (type.DeclaredAccessibility == Accessibility.Public && 
+                type.ContainingType == null &&
+                type.ContainingNamespace.IsGlobalNamespace)
+            {
+                context.ReportDiagnostic(type.CreateDiagnostic(Rule));
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/DeclareTypesInNamespaces.cs]: Hah. You're right, you _did_ cherry-pick an easy one :-)

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/DeclareTypesInNamespacesTests.cs [hunk 2]
```diff
@@ -10,14 +10,97 @@ namespace Microsoft.ApiDesignGuidelines.Analyzers.UnitTests
 {
     public class DeclareTypesInNamespacesTests : DiagnosticAnalyzerTestBase
     {
+        private static readonly string s_ruleId = DeclareTypesInNamespacesAnalyzer.RuleId;
+        private static readonly string s_message = MicrosoftApiDesignGuidelinesAnalyzersResources.DeclareTypesInNamespacesMessage;
+
+        [Fact]
+        public void OuterTypeInGlobalNamespace_Warns()
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/DeclareTypesInNamespacesTests.cs]: Huh. Now this is a nice convention that I haven't seen anybody else use: putting the C# and VB tests into the same test case. It does make it easy to see them together, and helps you not forget to implement the VB case for each test.

**Comment by lgolding** [comment]
:+1: with a caveat. Please take a look at the [FxCop implementation](http://index/#DesignRules/DeclareTypesInNamespaces.cs,3eba5ee7aa6ba974). It excludes certain ASP.NET types that are emitted into the global namespace, and it includes not only `public` types (`TargetVisibilities.ExternallyVisible`), but also types that are `TargetVisibilities.Overridable` -- and I don't know what that means for a type.

**Comment by srivatsn** [comment]
:+1:  modulo the special cases that Larry pointed out.

**Comment by nguerrera** [comment]
Regarding the special cases:

`TargetVisibilities.Overridable` in FxCop meant that the user could enable an option to run rules that normally only report issues on publicly visible API to instead analyze all API. I don't believe that option was ever exposed in VS nor do I think there is any equivalent for Roslyn analyzers. That leaves only `TargetVisibilities.ExternallyVisible`, which for named, non-nested types, DeclaredAccessibility == Public captures. 

I cannot reproduce the ASP.NET controls being emitted outside a namespace. No matter what I do in a web forms project (I've tried adding pages, forms, everything that extends System.Web.UI.Control is generated in the default namespace for my project. I can only guess that this issue was corrected at some point in VS and the special-casing in the rule is outdated. I therefore suggest that we leave out the special case from the port. We can add it back if someone hits it and has a repro.

**Comment by srivatsn** [comment]
Sounds good.

**Comment by lgolding** [comment]
:+1:

**Pull request merged**